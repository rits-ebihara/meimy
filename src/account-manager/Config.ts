/* eslint-disable @typescript-eslint/no-explicit-any */
import { IColorPalette } from '../IColorPalette';

const ACCOUNT_LIST = 'account_list';
const LAST_ACCOUNT = 'last_account';

/**
 * Configを設定するときの引数の型
 */
export interface ISetConfigParams {
    /**
     * モバイルアプリを一意に判別するID
     * @example 'jp.co.ricoh.jrits.eim.decisionformat'
     */
    appServiceId: string;
    /**
     * EIMアプリキーのプレフィックス
     * @description アプリ一覧画面で、ここで指定した文字列で始まるキーのアプリを表示する
     * @example 'DWS'
     */
    appKeyPrefix: string;
    /**
     * カラーパレット
     * @description http://paletton.com で作成した配色を指定します。Triad(3-colors)のみ使用可能。
     */
    colorPalets: IColorPalette;
    /**
     * 認証が終わった後のアプリのページ名
     * @description React Navigation でのページ名を指定します。
     */
    startPage: string;
    /**
     * Native Baseのテーマ
     * @description ```node node_modules/native-base/ejectTheme.js``` で出力したテーマを指定します。
     * 詳しくは、https://qiita.com/keneo/items/a03d8f28ceb584bc3ff6 を参照
     */
    theme: any;
}

export class Config {
    private _appKeyPrefix: string;
    private _appServiceId: string;
    private _colorPalets: IColorPalette;
    private _theme: any;
    private _startPage: string;
    public get appKeyPrefix() { return this._appKeyPrefix; }
    public get appServiceId() { return this._appServiceId; }
    public get colorPalets() { return this._colorPalets; }
    public get startPage() { return this._startPage; }
    public get theme() { return this._theme; }
    public get accountListServiceName() {
        return `${this._appServiceId}.${ACCOUNT_LIST}`;
    }
    public get lastAccountServiceName() {
        return `${this._appServiceId}.${LAST_ACCOUNT}`;
    }
    /*
$color-primary-0: #2F4172;	// Main Primary color
$color-primary-1: #7986AC;
$color-primary-2: #4F608F;
$color-primary-3: #162756;
$color-primary-4: #061439;

$color-secondary-1-0: #AA9839;	// Main Secondary color (1)
$color-secondary-1-1: #FFF1AA;
$color-secondary-1-2: #D4C36A;
$color-secondary-1-3: #806E15;
$color-secondary-1-4: #554700;

$color-secondary-2-0: #AA6E39;	// Main Secondary color (2)
$color-secondary-2-1: #FFD2AA;
$color-secondary-2-2: #D49C6A;
$color-secondary-2-3: #804715;
$color-secondary-2-4: #552800;

    */
    public constructor() {
        this._appServiceId = '';
        this._appKeyPrefix = '';
        this._startPage = '';
        this._theme = {};
        this._colorPalets = {
            $colorPrimary0: '#2F4172',
            $colorPrimary1: '#7986AC',
            $colorPrimary2: '#4F608F',
            $colorPrimary3: '#162756',
            $colorPrimary4: '#061439',
            $colorSecondary1p0: '#AA9839',
            $colorSecondary1p1: '#FFF1AA',
            $colorSecondary1p2: '#D4C36A',
            $colorSecondary1p3: '#806E15',
            $colorSecondary1p4: '#554700',
            $colorSecondary2p0: '#AA6E39',
            $colorSecondary2p1: '#FFD2AA',
            $colorSecondary2p2: '#D49C6A',
            $colorSecondary2p3: '#804715',
            $colorSecondary2p4: '#552800',
            $frontColor: '#333',
            $frontDisabledColor: '#999',
            $invertColor: '#fff',
        };
    }
    public setConfig(params: ISetConfigParams) {
        this._appKeyPrefix = params.appKeyPrefix;
        this._appServiceId = params.appServiceId;
        this._colorPalets = params.colorPalets;
        this._startPage = params.startPage;
        this._theme = params.theme;
    }
}

const config = new Config();
export const getConfig = (userConfig?: Config) => {
    return userConfig || config;
};
