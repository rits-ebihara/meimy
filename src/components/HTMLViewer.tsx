import React, { Component } from 'react';
import { NativeSyntheticEvent, WebView, WebViewMessageEventData } from 'react-native';


interface IProps {
    html: string;
}

interface IState {
    height: number;
}

export class HtmlViewer extends Component<IProps, IState> {
    public constructor(props: IProps) {
        super(props);
        this.state = {
            height: 200,
        };
    }
    public render() {
        return (
            <WebView
                source={{ html: this.createHtml(this.props.html) }}
                style={{ height: this.state.height }}
                onMessage={this.onMessage}
            />
        );
    }
    private onMessage = (event: NativeSyntheticEvent<WebViewMessageEventData>) => {
        let height = Number(event.nativeEvent.data);
        if (!Number.isNaN(height)) {
            // height = height / PixelRatio.get();
            this.setState({
                height,
            });
        }
    }
    private createHtml = (html: string) => {
        const completedHtml = `<!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width,initial-scale=1">
            <style>
                html, body {
                    font-size: 14pt;
                }
                #container {
                    position: absolute;
                    top: 0;
                    left: 0;
                    right: 0;
                    margin: 0;
                    padding: 0;
                }
                p {
                    margin: 0;
                    padding: 0;
                }
            </style>
            <script>
                window.addEventListener("load", () => {
                    setTimeout(() => {
                        const container = document.getElementById('container');
                        window.ReactNativeWebView.postMessage(container.clientHeight);
                    }, 100);
                });
            </script>
        </head>
        <body>
            <div id="container">
                ${html}
            </div>
        </body>
        </html>`;
        return completedHtml;
    }
}

export default HtmlViewer
