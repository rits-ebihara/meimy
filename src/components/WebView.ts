import { Platform, WebView } from "react-native";
import WKWebView from 'react-native-wkwebview-reborn';

export default Platform.OS === 'ios' ? WKWebView : WebView;