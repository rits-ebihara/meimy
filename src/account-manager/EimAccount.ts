import { getGenericPassword, setGenericPassword } from 'react-native-keychain';

import { IUserDoc } from '../eim-service/EIMDocInterface';
import { EIMServiceAdapter } from '../eim-service/EIMServiceAdapter';
import { getConfig } from './Config';
import { IEimAccount, IEimAccountBase } from './IEimAccount';

const config = getConfig();

export class EimAccount implements IEimAccount {
    public appKey: string;
    public domain: string;
    public eimTokens: string[];
    public user?: IUserDoc;
    public siteName: string;
    private serviceAdapter: EIMServiceAdapter | null;
    public constructor() {
        this.appKey = '';
        this.domain = '';
        this.eimTokens = [];
        this.serviceAdapter = null;
        this.siteName = '';
    }
    public load = async () => {
        const lastAccountString = await getGenericPassword({ service: config.lastAccountServiceName });
        if (!lastAccountString) {
            return null;
        }
        if (typeof lastAccountString === 'boolean') {
            return null;
        } else {
            return JSON.parse(lastAccountString.password) as IEimAccountBase;
        }
    }
    public save = async () => {
        const lastAccountString = JSON.stringify(this);
        await setGenericPassword('dummy', lastAccountString, { service: config.lastAccountServiceName });
    }
    public loadUser = async () => {
        const sa = new EIMServiceAdapter(this.domain);
        const user = await sa.getLoginUser(this.eimTokens);
        if (!!user) {
            this.user = user;
        }
    }
    public getServiceAdapter = () => {
        if (!this.serviceAdapter) {
            this.serviceAdapter = new EIMServiceAdapter(this.domain);
        }
        return this.serviceAdapter;
    }
    public getDepartmentName() {
        const user = this.user;
        const profile = user ? user.properties.profile : null;
        const department = profile ? profile.department : null;
        const properties = department ? department.properties : null;
        const fullLabel = properties ? properties.fullLabel || null : null;
        return fullLabel;
    }

    public clear = () => {
        this.appKey = '';
        this.domain = '';
        this.eimTokens = [];
        this.serviceAdapter = null;
        this.user = undefined;
    }
}

const eimAccount = new EimAccount();

export const getEimAccount = (obj?: EimAccount) => {
    return obj || eimAccount;
};

// eslint-disable-next-line max-len
// adb shell am start -a android.intent.action.VIEW "eimapplink-decision-form://decision-form/?link=https%3A%2F%2Fapp-dev54.ope.azure.ricoh-eim.com%2F\&hash=%2Fapps%2FDWS_mobile_r%2Fdocuments%2F11b779a79e6d42adab15f0fd83f58b34"
// eslint-disable-next-line max-len
// https://applink.eim.ricoh.com/applink/decision-form?link=https%3A%2F%2F{eim domain}%2F&hash=%2Fapps%2F{app id}%2Fdocuments%2F{doc id}&andid=jp.co.ricoh.jrits.eim.decisionformat&iosid=jp.co.ricoh.jrits.eim.decisionformat
// eslint-disable-next-line max-len
// https://applink.eim.ricoh.com/applink/decision-form?link=https%3A%2F%2Fapp-dev54.ope.azure.ricoh-eim.com%2F&hash=%2Fapps%2FDWS%2Fdocuments%2F95f6a02d6d7a448dadeefb2819ea4e2f&andid=jp.co.ricoh.jrits.eim.decisionformat&iosid=jp.co.ricoh.jrits.eim.decisionformat
