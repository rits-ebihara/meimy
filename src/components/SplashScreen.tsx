import { Container, Text, Toast, View } from 'native-base';
import React, { Component } from 'react';
import { AppState, Linking, TextStyle, ViewStyle } from 'react-native';
import DevInfo from 'react-native-device-info';
import UrlParse from 'url-parse';

import { config, getEimAccount } from '../account-manager';
import navigateController from '../account-manager/actions/NavigateActions';
import { ICombinedNavProps } from '../redux-helper/redux-helper';

let runOnLink = false;
export interface INavProps {
    parameter: string;
}

export interface ISplashState {
    appName: string;
}
//#region Styles
export abstract class SplashScreen<T extends ISplashState> extends Component<ICombinedNavProps<T>> {
    public render() {
        const { theme } = config;
        const containerStyle: ViewStyle = {
            alignItems: 'center',
            backgroundColor: theme.brandPrimary,
            justifyContent: 'center',
        };

        const appNameStyle: TextStyle = {
            color: theme.inverseTextColor,
            fontSize: 32,
        };
        const versionStyle: TextStyle = {
            color: theme.inverseTextColor,
            fontSize: 20,
            position: 'absolute',
            top: 0,
            alignSelf: 'flex-end',
        };
        return (
            <Container style={containerStyle}>
                <Text style={versionStyle}>ver.{DevInfo.getVersion()}</Text>
                <View>
                    <Text style={appNameStyle}>{this.props.state.appName}</Text>
                </View>
            </Container>
        );
    }
    public componentDidMount = () => {
        Linking.getInitialURL().then(this.linkInitialURL);
        Linking.addEventListener('url', this.urlEvent);
        this.checkConnectOnResume();
    }
    private checkConnectOnResume = () => {
        const eimAccount = getEimAccount();
        AppState.addEventListener('change', async (state) => {
            if (state === 'active') {
                const token = eimAccount.eimTokens;
                if (token.length !== 0) {
                    const result = await eimAccount.getServiceAdapter().validateToken(token);
                    if (!result) {
                        eimAccount.eimTokens = [];
                        Toast.show({
                            text: 'サーバーとの接続が切れました。\n再ログインしてください。',
                        });
                        navigateController.openAccountManager(this.props.navigation, this.props.dispatch);
                    }
                }
            }
        });
    }
    private linkInitialURL = async (url: string | null) => {
        // ディープリンクから起動された場合
        // 通常起動の場合も反応する。その場合 url = null となる。
        if (!url || runOnLink) {
            // スプラッシュを表示するため、インターバルを取る
            setTimeout(() => {
                navigateController.openAccountManager(this.props.navigation, this.props.dispatch);
            }, 500);
            return;
        }
        runOnLink = true;
        const receiveUrl = UrlParse(url, '', true);
        const { query } = receiveUrl;
        if (!!query.link && query.hash) {
            navigateController.openAccountManager(this.props.navigation,
                this.props.dispatch,
                query.link,
                query.hash);
        }
    }
    private urlEvent = async (e: { url: string }) => {
        const urlString = e.url;
        const receiveUrl = UrlParse(urlString, '', true);
        const { query } = receiveUrl;
        if (!query) { return; }
        // スプラッシュ画面を表示するため、インターバルを置く
        if (this.didComeFromAccountManager(
            receiveUrl.hostname, query)) {
            await this.fromAccountManager(
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                query.token!, query.domain!, query.appkey!,
                query.siteName, query.link,
            );
        } else if (!!query.link
            && query.hash) {
            // メール等文書リンクから来た場合
            navigateController.openAccountManager(
                this.props.navigation, this.props.dispatch,
                query.link, query.hash);
        }
    }
    private didComeFromAccountManager = (hostName: string, query: { [key: string]: string | undefined }) => {
        return (hostName === 'authed'
            && !!query.domain
            && !!query.appkey
            && !!query.token);
    }
    private fromAccountManager = async (
        token: string, domain: string,
        appKey: string, siteName: string | undefined,
        link: string | undefined,
    ) => {
        const eimAccount = getEimAccount();
        eimAccount.eimTokens = token.split(',');
        eimAccount.domain = domain;
        eimAccount.appKey = appKey;
        eimAccount.siteName = siteName || domain;
        eimAccount.user = undefined;
        eimAccount.save();
        await eimAccount.loadUser();
        if (!!link) {
            const linkUrl = UrlParse(link);
            const hashes = linkUrl.hash.split('/');
            if (hashes.length === 5) {
                const navProps: INavProps = {
                    parameter: hashes[4],
                };
                this.props.navigation.navigate(config.startPage, navProps);
            }
            return;
        } else {
            this.props.navigation.navigate(config.startPage);
        }
    }
}
// eslint-disable-next-line max-len
// adb shell am start -a android.intent.action.VIEW "eimapplink-apptest1://apptest1/?link=https%3A%2F%2Fapp-dev54.ope.azure.ricoh-eim.com%2F\&hash=%2Fapps%2FDWS%2Fdocuments%2Fa7c2cee5e03a42538e90bec4df8789d8"
