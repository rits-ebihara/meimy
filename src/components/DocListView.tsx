import Clone from 'clone';
import { Button, Content, List, Spinner, Text } from 'native-base';
import React, { Component } from 'react';
import { RefreshControl } from 'react-native';

import { getEimAccount } from '../account-manager/EimAccount';
import { IDocListForView, IDocListRowForView } from '../eim-service/IDocListForView';
import { IDocListSearchOption, IDocListSort, SearchCondition } from '../eim-service/IDocListSearchOption';

export type CreateRowElement<T>
    = (row: IDocListRowForView<T>, cols: T) => JSX.Element;
export interface IDocListViewProps<T> {
    appKey?: string;
    docListKey: string;
    rowCountAtOnce: number;
    rowElement: CreateRowElement<T>;
    searchCondition?: SearchCondition<T>[];
    sortCondition?: IDocListSort<T>[];
    theme: { brandPrimary: string; textColor: string };
    hide?: boolean;
    onFinishLoad?: () => void;
}

export interface IDocListViewLocalState<T> {
    docListData?: IDocListForView<T>;
    offset: number;
    onSearch: boolean;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class DocListView<T = any> extends Component<IDocListViewProps<T>, IDocListViewLocalState<T>> {
    private $isMounted: boolean = false;
    public constructor(props: IDocListViewProps<T>) {
        super(props);
        this.state = {
            offset: 0,
            onSearch: false,
        };
    }
    public render = () => {
        const { docListData } = this.state;
        const list = this.createRowElement(docListData);
        const moreButton = this.createMoreButton();
        const refreshControl = <RefreshControl
            refreshing={this.state.onSearch}
            onRefresh={this.reload}
        />;
        return (
            <Content refreshControl={refreshControl}
                style={{ display: !!this.props.hide ? 'none' : 'flex' }}>
                <List>
                    {list ?
                        list :
                        <Spinner color={this.props.theme.brandPrimary} />}
                </List>
                {moreButton}
                {!!docListData && docListData.docList.length === 0
                    ? <Text style={{ color: this.props.theme.textColor }}>
                        対象の文書はありません。</Text>
                    : null}
            </Content>
        );
    }
    public componentDidMount = () => {
        this.$isMounted = true;
        // iOS では、setTimeout で setState しないとフリーズしてしまう
        setTimeout(this.loadDocList.bind(this, 0), 100);
    }
    public componentWillUnmount = () => {
        this.$isMounted = false;
    }
    public reload = () => {
        this.loadDocList(0);
    }
    private callLoadDocListData = async (props: IDocListViewProps<T>, offset: number) => {
        const searchOptions: IDocListSearchOption<T> = {
            limit: props.rowCountAtOnce,
            offset,
            search: props.searchCondition,
            sort: props.sortCondition,
        };
        const eimAccount = getEimAccount();
        return await eimAccount.getServiceAdapter().getDocListForView<T>(
            eimAccount.eimTokens,
            props.appKey || eimAccount.appKey,
            props.docListKey,
            searchOptions,
        );
    }

    private createRowElement = (docListData: IDocListForView<T> | undefined) => {
        return !docListData ? null :
            docListData.docList.map((row) => {
                const cols: {
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    [key: string]: any;
                } = {};
                row.columnValues.forEach((col) => {
                    cols[col.propertyName as string] = col.value;
                });
                return this.props.rowElement(row, cols as T);
            });
    }

    private createMoreButton = () => {
        const { docListData } = this.state;
        if (!docListData) { return null; }
        if (docListData.metrics.totalCount <= this.state.offset) {
            return null;
        }
        return ((this.state.onSearch) ?
            <Spinner color={this.props.theme.brandPrimary} />
            : <Button full light
                onPress={this.loadDocList.bind(this, this.state.offset)}>
                <Text>さらに表示</Text>
            </Button>);
    }

    private loadDocList = async (offset: number) => {
        if (offset === 0) {
            this.setState({
                docListData: undefined,
                onSearch: true,
            });

        } else {
            this.setState({
                onSearch: true,
            });
        }
        return this.callLoadDocListData(this.props, offset).then((result) => {
            const { state } = this;
            const docListData = Clone(result);
            if (!!state.docListData) {
                docListData.docList =
                    state.docListData.docList.concat(docListData.docList);
            }
            if (this.$isMounted) {
                this.setState({
                    docListData,
                    offset: offset + this.props.rowCountAtOnce,
                    onSearch: false,
                });
            }
            if (this.props.onFinishLoad) { this.props.onFinishLoad(); }
        }).catch(((e) => {
            this.setState({
                onSearch: false,
            });
            console.warn(e);
        }));
    }
}
