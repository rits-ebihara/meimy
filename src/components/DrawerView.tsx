import { Button, Icon, Text, Thumbnail, View } from 'native-base';
import React, { Component } from 'react';
import { ImageURISource, ViewStyle } from 'react-native';
import { DrawerItemsProps } from 'react-navigation';

import { getEimAccount } from '../account-manager/EimAccount';
import dummyAvatar from '../resources/user.png';

// import { app, routePageNames } from '../../Commons';
// import theme, { colorPallet } from '../Styles';

export interface IDrawerViewProps extends DrawerItemsProps {
    borderColor: string;
    buttonColor: string;
    appDisplayName: string;
    appVersion: string;
    splashPageName: string;
}

export interface ILocalState {
    avatarFaceUrl?: ImageURISource;
}

export class DrawerContent extends Component<IDrawerViewProps, ILocalState> {
    private $isMounted = false;
    public constructor(props: IDrawerViewProps) {
        super(props);
        this.state = {};
    }
    public render() {
        const viewStyle: ViewStyle = {
            borderBottomColor: this.props.borderColor,
            borderBottomWidth: 1,
            borderStyle: 'solid',
            padding: 10,
        };
        const rowStyle: ViewStyle = Object.assign<{}, ViewStyle, ViewStyle>({}, viewStyle, {
            flexDirection: 'row',
        });
        const colStyle: ViewStyle = Object.assign<{}, ViewStyle, ViewStyle>({}, viewStyle, {
            flexDirection: 'column',
        });
        const eimAccount = getEimAccount();
        if (!eimAccount.user) {
            return null;
        }
        const { user } = eimAccount;
        return (
            <View>
                <View style={{ flexDirection: 'row-reverse' }}>
                    <Button
                        key="closeButton"
                        transparent
                        onPress={() => this.props.navigation.toggleDrawer()}>
                        <Text>閉じる</Text>
                    </Button>
                </View>
                <View style={rowStyle}>
                    <View style={{ width: 88 }}>
                        <Thumbnail circular large source={this.state.avatarFaceUrl || dummyAvatar} />
                    </View>
                    <View style={{ width: 180 }}>
                        <Text>{user.properties.displayName}</Text>
                        <Text>{eimAccount.getDepartmentName()}</Text>
                    </View>
                </View>
                <View style={colStyle}>
                    <Text>{}</Text>
                    <Text note>{eimAccount.domain}</Text>
                    <Button
                        bordered color={this.props.buttonColor} iconLeft rounded
                        style={{ alignSelf: 'center', margin: 10 }}
                        onPress={this.onPressChangeSite.bind(this)}>
                        <Icon name="md-swap" />
                        <Text>接続先の切り替え</Text>
                    </Button>
                </View>
                <View style={colStyle} >
                    <Text note>このアプリについて</Text>
                    <Text style={{
                        fontSize: 16,
                        fontWeight: 'bold',
                    }}>{this.props.appDisplayName}</Text>
                    <Text style={{ alignSelf: 'flex-end' }}>ver.{this.props.appVersion}</Text>
                </View>
            </View>
        );
    }
    public componentDidMount = async () => {
        this.$isMounted = true;
        const eimAccount = getEimAccount();
        const { user } = eimAccount;
        const faceImage = user ? user.properties.faceImage : null;
        if (!faceImage) {
            return;
        }
        return eimAccount.getServiceAdapter()
            .getAttachmentFile(eimAccount.eimTokens, faceImage)
            .then((result) => {
                if (!this.$isMounted) { return; }
                this.setState({
                    avatarFaceUrl: {
                        uri: result.url,
                    },
                });
            });
    }
    public componentWillUnmount = () => {
        this.$isMounted = false;
    }
    public onPressChangeSite = async () => {
        const eimAccount = getEimAccount();
        eimAccount.clear();
        await eimAccount.save();
        this.props.navigation.navigate(this.props.splashPageName);
    }

}
