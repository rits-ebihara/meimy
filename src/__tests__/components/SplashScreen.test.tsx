import Enzyme, { shallow, ShallowWrapper } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { Toast } from 'native-base';
import React from 'react';
import { AppState, Linking } from 'react-native';
import { mocked } from 'ts-jest/utils';

import navigateController from '../../account-manager/actions/NavigateActions';
import { getEimAccount } from '../../account-manager/EimAccount';
import { ISplashState, SplashScreen } from '../../components/SplashScreen';
import { ICombinedNavProps } from '../../redux-helper/redux-helper';
import { sleep } from '../commonFunctions';

Enzyme.configure({ adapter: new EnzymeAdapter() });

Enzyme.configure({ adapter: new EnzymeAdapter() });

const createEimAccount = () => ({
    user: {
        properties: {
            displayName: 'ricoh-00001',
            faceImage: null as string | null,
        },
    },
    getDepartmentName: jest.fn().mockReturnValue('sosiki-full'),
    getServiceAdapter: jest.fn().mockImplementation(() => ({
        validateToken: jest.fn(),
    })),
    clear: jest.fn(),
    save: jest.fn(),
    domain: 'site-domain',
    eimTokens: ['token'],
    loadUser: jest.fn(),
});

jest.mock('../../account-manager/EimAccount', () => ({
    getEimAccount: jest.fn(),
}));

jest.mock('../../account-manager/actions/NavigateActions', () => ({
    openAccountManager: jest.fn(),
}));

jest.mock('native-base', () => ({
    ...jest.requireActual('native-base'),
    Toast: {
        show: jest.fn(),
    }
}));

jest.mock('AppState', () => ({
    addEventListener: jest.fn(),
}));

jest.mock('Linking', () => ({
    getInitialURL: jest.fn().mockResolvedValue(null),
    addEventListener: jest.fn(),
}));

jest.mock('react-native-device-info', () => ({
    getVersion: jest.fn().mockReturnValue('1.0.0'),
}));

jest.mock('../../account-manager/Config', () => ({
    getConfig: jest.fn().mockReturnValue({
        startPage: 'start-page',
        theme: {
            brandPrimary: '#009',
            inverseTextColor: '#fff',
        },
    }),
}));

const createProps = () => ({
    state: {
        appName: 'test-app-name',
    },
    navigation: {
        navigate: jest.fn(),
    },
    dispatch: jest.fn(),
});

describe('render', () => {
    test('init', () => {
        const props = createProps();
        const wrapper = shallow(<SplashScreen {...props as any} />, { disableLifecycleMethods: true });
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('event', () => {
    let props: any;
    let wrapper: ShallowWrapper<Readonly<ICombinedNavProps<ISplashState>> & Readonly<{
        children?: React.ReactNode;
    }>, Readonly<{}>, SplashScreen<ISplashState>>;
    let instance: SplashScreen<ISplashState>;
    beforeEach(() => {
        mocked(getEimAccount).mockClear();
        mocked(Toast.show).mockClear();
        mocked(navigateController.openAccountManager).mockClear();
        props = createProps();
        wrapper = shallow<SplashScreen<ISplashState>>(
            <SplashScreen {...props} />, { disableLifecycleMethods: true });
        instance = wrapper.instance();
    });
    test('componentDidMout', async () => {
        instance['linkInitialURL'] = jest.fn();
        await instance.componentDidMount();
        expect(Linking.getInitialURL).toBeCalled();
        expect(Linking.addEventListener).toBeCalledWith('url', instance['urlEvent']);
        expect(AppState.addEventListener).toBeCalledWith(
            'change', instance['reloginAnnounce']);
        expect(instance['linkInitialURL']).toBeCalledWith(null);
    });
    describe('reloginAnnounce', () => {
        test('background', async () => {
            const eimAccount = createEimAccount();
            mocked(getEimAccount).mockReturnValue(eimAccount as any);
            await instance['reloginAnnounce']('background');
            expect(eimAccount.getServiceAdapter().validateToken).not.toBeCalled();
        });
        test('active/invalid token', async () => {
            const eimAccount = createEimAccount();
            mocked(eimAccount.getServiceAdapter).mockReturnValue(({
                validateToken: jest.fn().mockReturnValue(false),
            }));
            mocked(getEimAccount).mockReturnValue(eimAccount as any);
            await instance['reloginAnnounce']('active');
            expect(eimAccount.getServiceAdapter().validateToken).toBeCalled();
            expect(Toast.show).toBeCalledWith({
                text: 'サーバーとの接続が切れました。\n再ログインしてください。'
            });
            expect(navigateController.openAccountManager).toBeCalledWith(
                props.navigation, props.dispatch,
            );
        });
        test('active/valid token', async () => {
            const eimAccount = createEimAccount();
            mocked(eimAccount.getServiceAdapter).mockReturnValue(({
                validateToken: jest.fn().mockReturnValue(true),
            }));
            mocked(getEimAccount).mockReturnValue(eimAccount as any);
            await instance['reloginAnnounce']('active');
            expect(eimAccount.getServiceAdapter().validateToken).toBeCalled();
            expect(Toast.show).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();
        });
    });
    describe('linkInitialURL', () => {
        test('url is null | runOnLink = true', async () => {
            instance['runOnLink'] = false;
            await instance['linkInitialURL'](null);
            await sleep();
            expect(navigateController.openAccountManager).toBeCalledWith(
                props.navigation, props.dispatch
            );
            expect(instance['runOnLink']).toEqual(false);
            // runOnLink = true
            instance['runOnLink'] = true;
            await instance['linkInitialURL']('http://exsample.com/');
            await sleep();
            expect(navigateController.openAccountManager).toBeCalledWith(
                props.navigation, props.dispatch
            );
        });
        test('url is not null && no link || no hash', async () => {
            instance['runOnLink'] = false;
            await instance['linkInitialURL']('http://exsample.com/');
            expect(navigateController.openAccountManager).not.toBeCalled();
            await instance['linkInitialURL']('http://exsample.com/?link=hoge');
            expect(navigateController.openAccountManager).not.toBeCalled();
            await instance['linkInitialURL']('http://exsample.com/?hash=hoge');
            expect(navigateController.openAccountManager).not.toBeCalled();
        });
        test('url is not null && link and hash is not null', async () => {
            instance['runOnLink'] = false;
            await instance['linkInitialURL']('http://exsample.com/?link=hoge&hash=fuga');
            expect(navigateController.openAccountManager).toBeCalledWith(
                props.navigation, props.dispatch,
                'hoge', 'fuga'
            );
        });
    });
    describe('urlEvent', async () => {
        test('call from mail - link, hash with or without', async () => {
            instance['fromAccountManager'] = jest.fn();
            await instance['urlEvent']({ url: 'https://exsample.com/' });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();
            await instance['urlEvent']({ url: 'https://exsample.com/?hash=hoge' });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();
            await instance['urlEvent']({ url: 'https://exsample.com/?link=hoge' });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();
            await instance['urlEvent']({ url: 'https://exsample.com/?hash=hoge&link=fuga' });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).toBeCalledWith(
                props.navigation, props.dispatch, 'fuga', 'hoge'
            );
        });
        test('call from account manager', async () => {
            instance['fromAccountManager'] = jest.fn();
            await instance['urlEvent']({ url: 'https://authed/' });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();

            await instance['urlEvent']({
                url: 'https://authed/?appkey=hoge&link=fuga&siteName=uga&link=quga'
            });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();

            await instance['urlEvent']({
                url: 'https://authed/?domain=hoge&link=fuga&siteName=uga&link=quga'
            });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();

            await instance['urlEvent']({
                url: 'https://authed/?domain=hoge&appkey=fuga&siteName=uga&link=quga'
            });
            expect(instance['fromAccountManager']).not.toBeCalled();
            expect(navigateController.openAccountManager).not.toBeCalled();

            await instance['urlEvent']({
                url: 'https://authed/?domain=hoge&appkey=fuga&token=toga&siteName=uga&link=quga'
            });
            expect(navigateController.openAccountManager).not.toBeCalled();
            expect(instance['fromAccountManager']).toBeCalledWith(
                'toga', 'hoge', 'fuga', 'uga', 'quga'
            );
        });
    });
    describe('fromAccountManager', () => {
        test('no link', async () => {
            const eimAccount = createEimAccount();
            mocked(getEimAccount).mockReturnValue(eimAccount as any);

            await instance['fromAccountManager'](
                'token', 'domain', 'app-key', 'site-name', undefined
            );

        });
    });
});
