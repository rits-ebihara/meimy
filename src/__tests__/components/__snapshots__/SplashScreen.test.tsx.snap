// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render init 1`] = `
<Styled(Container)
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#009",
      "justifyContent": "center",
    }
  }
>
  <Styled(Text)
    style={
      Object {
        "alignSelf": "flex-end",
        "color": "#fff",
        "fontSize": 20,
        "position": "absolute",
        "top": 0,
      }
    }
  >
    ver.
    1.0.0
  </Styled(Text)>
  <Styled(ViewNB)>
    <Styled(Text)
      style={
        Object {
          "color": "#fff",
          "fontSize": 32,
        }
      }
    >
      test-app-name
    </Styled(Text)>
  </Styled(ViewNB)>
</Styled(Container)>
`;
